### PR TITLE
add reverberated TDNN+LSTM recipe on AMI; a fix to aspire recipe

### PR DESCRIPTION
**Pull request by tomkocse**

### CODE REVIEW

**Comment by danpovey** [comment]
@vijayaditya, do you have time to review this?

**Comment by vijayaditya** [comment]
Will do.

Vijay

File: egs/ami/s5b/RESULTS_mdm [hunk 1]
```diff
@@ -86,6 +86,13 @@
...
 #  cleanup + chain TDNN+LSTM model, MDM audio and alignments from IHM data
-# *** best system ***
 %WER 34.6 | 15116 94508 | 69.6 17.6 12.9 4.1 34.6 62.3 | 0.687 | exp/mdm8/chain_cleaned/tdnn_lstm1i_sp_bi_ihmali_ld5/decode_dev/ascore_9/dev_hires_o4.ctm.filt.sys
 %WER 37.1 | 14343 90002 | 66.3 18.8 14.9 3.4 37.1 62.3 | 0.659 | exp/mdm8/chain_cleaned/tdnn_lstm1i_sp_bi_ihmali_ld5/decode_eval/ascore_9/eval_hires_o4.ctm.filt.sys
+
+
+# local/chain/multi_condition/run_tdnn_lstm.sh --mic mdm8 --use-ihm-ali true --train-set train_cleaned --gmm tri3_cleaned
+# cleanup + chain TDNN+LSTM model, MDM original + IHM reverberated data, alignments from IHM data
+# *** best system ***
+%WER 31.8 | 14488 94497 | 71.8 15.4 12.8 3.5 31.8 62.7 | 0.698 | exp/mdm8/chain_cleaned_rvb/tdnn_lstm1i_sp_rvb_bi_ihmali/decode_dev/ascore_10/dev_hires_o4.ctm.filt.sys
```
Comment (vijayaditya) [inline egs/ami/s5b/RESULTS_mdm]: @tomkocse Very impressive results. Please add the version number of your script which generates these results (eg. run_tdnn_lstm_1a.sh ) as we might replace run_tdnn_lstm.sh with a better tdnn_lstm architecture in the future.

Move this script to tuning directory as run_tdnn_lstm_1b.sh and softlink it from run_tdnn_lstm.sh

File: egs/ami/s5b/RESULTS_sdm [hunk 2]
```diff
@@ -79,16 +79,22 @@
...
 
 # local/chain/tuning/run_tdnn_lstm_1i.sh --mic sdm1 --use-ihm-ali true --train-set train_cleaned  --gmm tri3_cleaned
-# cleanup + chain TDNN model, SDM audio + alignments from ihm data.
-# *** best system ***
+# cleanup + chain TDNN+LSTM model, SDM audio + alignments from ihm data.
 %WER 37.6 | 15122 94495 | 66.1 18.7 15.1 3.7 37.6 63.2 | 0.646 | exp/sdm1/chain_cleaned/tdnn_lstm1i_sp_bi_ihmali_ld5/decode_dev/ascore_10/dev_hires_o4.ctm.filt.sys
 %WER 40.9 | 13807 89961 | 62.4 20.0 17.6 3.3 40.9 65.7 | 0.612 | exp/sdm1/chain_cleaned/tdnn_lstm1i_sp_bi_ihmali_ld5/decode_eval/ascore_10/eval_hires_o4.ctm.filt.sys
+
+
+# local/chain/multi_condition/run_tdnn_lstm.sh --mic sdm1 --use-ihm-ali true --train-set train_cleaned --gmm tri3_cleaned
```
Comment (vijayaditya) [inline egs/ami/s5b/RESULTS_sdm]: same comment as above; start versioning your run*.sh scripts.

File: egs/ami/s5b/local/chain/multi_condition/run_tdnn_lstm.sh [hunk 3]
```diff
@@ -0,0 +1,334 @@
+#!/bin/bash
+
+# This is a chain-training script with TDNN+LSTM neural networks.
```
Comment (vijayaditya) [inline egs/ami/s5b/local/chain/multi_condition/run_tdnn_lstm.sh]: Not checking this in detail as I assume it is similar to multi_condition/run_tdnn.sh except for the xconfig and some lstm specific parameters.

File: egs/ami/s5b/local/chain/multi_condition/run_tdnn_lstm.sh [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+  fixed-affine-layer name=lda input=Append(-1,0,1,ReplaceIndex(ivector, t, 0)) affine-transform-file=$dir/configs/lda.mat
+
+  # the first splicing is moved before the lda layer, so no splicing here
+  relu-renorm-layer name=tdnn1 dim=1024
+  relu-renorm-layer name=tdnn2 input=Append(-1,0,1) dim=1024
+  relu-renorm-layer name=tdnn3 input=Append(-1,0,1) dim=1024
+
+  # check steps/libs/nnet3/xconfig/lstm.py for the other options and defaults
+  lstmp-layer name=lstm1 cell-dim=1024 recurrent-projection-dim=256 non-recurrent-projection-dim=256 delay=-3
+  relu-renorm-layer name=tdnn4 input=Append(-3,0,3) dim=1024
```
Comment (vijayaditya) [inline egs/ami/s5b/local/chain/multi_condition/run_tdnn_lstm.sh]: Try increasing the parameters further as you have more data than the SDM/MDM only case. Though we might not ultimately recommend a large parameter model it is better to know the least possible WER with this architecture and simulated multi_condition data.  You can try increasing the relu-norm-layer dim in increments of 256 or 128.

File: egs/ami/s5b/local/chain/multi_condition/run_tdnn_lstm.sh [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+    --chain.apply-deriv-weights false \
+    --chain.lm-opts="--num-extra-lm-states=2000" \
+    --egs.dir "$common_egs_dir" \
+    --egs.opts "--frames-overlap-per-eg 0" \
+    --egs.chunk-width $chunk_width \
+    --egs.chunk-left-context $chunk_left_context \
+    --egs.chunk-right-context $chunk_right_context \
+    --trainer.num-chunk-per-minibatch 64 \
+    --trainer.frames-per-iter 1500000 \
+    --trainer.num-epochs 4 \
```
Comment (vijayaditya) [inline egs/ami/s5b/local/chain/multi_condition/run_tdnn_lstm.sh]: Check if you can increase the number of epochs.

File: egs/aspire/s5/local/chain/tuning/run_tdnn_7b.sh [hunk 6]
```diff
@@ -100,6 +100,7 @@ if [ $stage -le 9 ]; then
     data/train data/train_temp_for_lats
   utils/data/combine_short_segments.sh \
       data/train_temp_for_lats $min_seg_len data/train_min${min_seg_len}
+  steps/compute_cmvn_stats.sh data/train_min${min_seg_len} exp/make_mfcc/train_min${min_seg_len} mfcc || exit 1;
```
Comment (vijayaditya) [inline egs/aspire/s5/local/chain/tuning/run_tdnn_7b.sh]: I recommend using the new style of not specifying logdirs.

**Comment by danpovey** [comment]
Vijay,  wouldn't the training be rather slow with more epochs?  IIRC he's
already doubled the amount of data versus the regular 3x-perturbed data, so
that affects the meaning of epochs.

**Comment by vijayaditya** [comment]
@danpovey Thanks for pointing out I didn't observe that the num-epochs was not reduced. @tomkocse Could you please generate the log-probability plots for the training run with multi_condition data. You might know if your training has converged.

**Pull request merged**